File: benchmarks/runtime/spring-boot/src/main/java/com/example/benchmark/controller/PostApiController.java
package com.example.benchmark.controller;

import com.example.benchmark.model.ApiToken;
import com.example.benchmark.model.Post;
import com.example.benchmark.repository.ApiTokenRepository;
import com.example.benchmark.repository.PostRepository;
import jakarta.validation.Valid;
import org.springframework.data.domain.PageRequest;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@RestController
@RequestMapping("/api/posts")
public class PostApiController {

    private final PostRepository posts;
    private final ApiTokenRepository tokens;

    public PostApiController(PostRepository posts, ApiTokenRepository tokens) {
        this.posts = posts;
        this.tokens = tokens;
    }

    @GetMapping
    public List<Post> list() {
        return posts.findAllByOrderByCreatedAtDesc(PageRequest.of(0, 50));
    }

    @GetMapping("/{id}")
    public ResponseEntity<Post> show(@PathVariable Long id) {
        return posts.findById(id)
            .map(ResponseEntity::ok)
            .orElse(ResponseEntity.notFound().build());
    }

    @PostMapping
    public ResponseEntity<?> create(@Valid @RequestBody Post body) {
        return ResponseEntity.status(HttpStatus.CREATED).body(posts.save(body));
    }

    @PatchMapping("/{id}")
    public ResponseEntity<Post> update(@PathVariable Long id, @RequestBody Post patch) {
        return posts.findById(id).map(p -> {
            if (patch.getTitle() != null) p.setTitle(patch.getTitle());
            if (patch.getBody() != null)  p.setBody(patch.getBody());
            if (patch.getAuthor() != null) p.setAuthor(patch.getAuthor());
            p.setPublished(patch.isPublished());
            return ResponseEntity.ok(posts.save(p));
        }).orElse(ResponseEntity.notFound().build());
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<Void> delete(@PathVariable Long id) {
        if (!posts.existsById(id)) return ResponseEntity.notFound().build();
        posts.deleteById(id);
        return ResponseEntity.noContent().build();
    }

    @GetMapping("/protected")
    public ResponseEntity<?> protectedStats(
        @RequestHeader(value = "Authorization", required = false) String authHeader
    ) {
        if (authHeader == null || !authHeader.startsWith("Bearer ")) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(Map.of("error", "missing or invalid Authorization header"));
        }
        String rawToken = authHeader.substring(7);
        Optional<ApiToken> token = tokens.findByToken(rawToken);
        if (token.isEmpty()) {
            return ResponseEntity.status(HttpStatus.UNAUTHORIZED)
                .body(Map.of("error", "invalid token"));
        }
        return ResponseEntity.ok(Map.of(
            "principal", token.get().getPrincipal(),
            "total_posts", posts.count()
        ));
    }
}

File: benchmarks/runtime/phoenix/lib/benchmark_web/controllers/post_api_controller.ex
defmodule BenchmarkWeb.PostApiController do
  use BenchmarkWeb, :controller
  alias Benchmark.{ApiToken, Post, Repo}
  import Ecto.Query

  def index(conn, _params) do
    posts = Post.recent(Repo)
    json(conn, posts)
  end

  def show(conn, %{"id" => id}) do
    case Repo.get(Post, id) do
      nil  -> conn |> put_status(404) |> json(%{error: "not found"})
      post -> json(conn, post)
    end
  end

  def create(conn, params) do
    changeset = Post.changeset(%Post{}, params)
    case Repo.insert(changeset) do
      {:ok, post}             -> conn |> put_status(201) |> json(post)
      {:error, changeset} ->
        errors = Ecto.Changeset.traverse_errors(changeset, fn {msg, _opts} -> msg end)
        conn |> put_status(422) |> json(%{error: inspect(errors)})
    end
  end

  def update(conn, %{"id" => id} = params) do
    case Repo.get(Post, id) do
      nil  -> conn |> put_status(404) |> json(%{error: "not found"})
      post ->
        changeset = Post.changeset(post, params)
        case Repo.update(changeset) do
          {:ok, updated}      -> json(conn, updated)
          {:error, changeset} ->
            errors = Ecto.Changeset.traverse_errors(changeset, fn {msg, _opts} -> msg end)
            conn |> put_status(422) |> json(%{error: inspect(errors)})
        end
    end
  end

  def delete(conn, %{"id" => id}) do
    case Repo.get(Post, id) do
      nil  -> conn |> put_status(404) |> json(%{error: "not found"})
      post ->
        Repo.delete!(post)
        send_resp(conn, 204, "")
    end
  end

  def protected(conn, _params) do
    case get_req_header(conn, "authorization") do
      ["Bearer " <> token | _] ->
        case ApiToken.verify(Repo, token) do
          {:ok, principal} ->
            total = Repo.one(from p in Post, select: count(p.id))
            json(conn, %{principal: principal, total_posts: total})
          {:error, _} ->
            conn |> put_status(401) |> json(%{error: "invalid token"})
        end
      _ ->
        conn |> put_status(401) |> json(%{error: "missing or invalid Authorization header"})
    end
  end
end
